packages/docker-engine/0001-Add-ability-to-read-default-OCI-Spec-from-a-file.patch

```diff
@@ -0,0 +1,212 @@
+ 	"os"
+@@ -24,6 +25,7 @@ import (
+ 	"github.com/docker/docker/pkg/stringid"
+ 	"github.com/docker/docker/rootless/specconv"
+ 	volumemounts "github.com/docker/docker/volume/mounts"
++	units "github.com/docker/go-units"
+ 	"github.com/moby/sys/mount"
+ 	"github.com/moby/sys/mountinfo"
+ 	"github.com/opencontainers/runc/libcontainer/apparmor"
+@@ -45,8 +47,13 @@ func WithRlimits(daemon *Daemon, c *container.Container) coci.SpecOpts {
+ 
+ 		// We want to leave the original HostConfig alone so make a copy here
+ 		hostConfig := *c.HostConfig
++
++		//Merge defaults in Base Spec file with Daemon defaults
++		daemon.mergeBaseSpecUlimits(s)
++
+ 		// Merge with the daemon defaults
+ 		daemon.mergeUlimits(&hostConfig)
++
+ 		for _, ul := range hostConfig.Ulimits {
+ 			rlimits = append(rlimits, specs.POSIXRlimit{
+ 				Type: "RLIMIT_" + strings.ToUpper(ul.Name),
+@@ -158,8 +165,12 @@ func WithApparmor(c *container.Container) coci.SpecOpts {
+ // WithCapabilities sets the container's capabilties
+ func WithCapabilities(c *container.Container) coci.SpecOpts {
+ 	return func(ctx context.Context, _ coci.Client, _ *containers.Container, s *coci.Spec) error {
++		defCaps := caps.DefaultCapabilities()
++		defCaps = append(defCaps, s.Process.Capabilities.Bounding...)
++		defCaps = getUniqueCapabilities(defCaps)
++
+ 		capabilities, err := caps.TweakCapabilities(
+-			caps.DefaultCapabilities(),
++			defCaps,
+ 			c.HostConfig.CapAdd,
+ 			c.HostConfig.CapDrop,
+ 			c.HostConfig.Privileged,
+@@ -171,6 +182,19 @@ func WithCapabilities(c *container.Container) coci.SpecOpts {
+ 	}
+ }
+ 
++func getUniqueCapabilities(defCaps []string) []string {
++	exists := make(map[string]bool, len(defCaps))
++	var uniqueCaps []string
++	for c := range defCaps {
++		if !exists[defCaps[c]] {
++			exists[defCaps[c]] = true
++			uniqueCaps = append(uniqueCaps, defCaps[c])
++		}
++	}
++
++	return uniqueCaps
++}
++
+ func resourcePath(c *container.Container, getPath func() (string, error)) (string, error) {
+ 	p, err := getPath()
+ 	if err != nil {
+@@ -1009,11 +1033,46 @@ func WithUser(c *container.Container) coci.SpecOpts {
+ 	}
+ }
+ 
++func loadDefaultOCISpec(filename string) (specs.Spec, error) {
++	var spec specs.Spec
++
++	baseSpec, errReadfile := os.ReadFile(filename)
++	// Fail when unable to read the base spec file
++	if errReadfile != nil {
++		return spec, fmt.Errorf("unable to read the base spec file provided in Daemon: %v", errReadfile)
++	}
++
++	errUnmarshall := json.Unmarshal(baseSpec, &spec)
++	if errUnmarshall != nil {
++		return spec, fmt.Errorf("unable to parse the base spec file provided in Daemon: : %s: %w", filename, errUnmarshall)
++	}
++
++	return spec, nil
++}
++
+ func (daemon *Daemon) createSpec(c *container.Container) (retSpec *specs.Spec, err error) {
+ 	var (
+ 		opts []coci.SpecOpts
+ 		s    = oci.DefaultSpec()
+ 	)
++
++	var filename string
++	runtime := daemon.configStore.GetRuntime(c.HostConfig.Runtime)
++	if runtime.BaseRuntimeSpec != "" {
```

Comment:
I think you should cache the parsed specs in `daemon.configStore`, otherwise every time a container is created the spec will have to be parsed and that adds additional overhead. This will be more aligned to what containerd does.